pkg/api/v1alpha1/cluster.go

```diff
@@ -433,12 +433,45 @@ func validateNetworking(clusterConfig *Cluster) error {
 	if err != nil {
 		return fmt.Errorf("invalid CIDR block for Services: %s. Please specify a valid CIDR block for service subnet", clusterConfig.Spec.ClusterNetwork.Services)
 	}
-	if clusterConfig.Spec.ClusterNetwork.CNI == "" {
-		return errors.New("cni not specified or empty")
+
+	return validateCNIPlugin(clusterConfig.Spec.ClusterNetwork)
+}
+
+func validateCNIPlugin(network ClusterNetwork) error {
+	if network.CNI != "" {
+		if network.CNIConfig != nil {
+			return fmt.Errorf("invalid format for cni plugin: both old and new formats used, use only the CNIConfig field")
+		}
+		logger.Info("Warning: CNI field is deprecated. Provide CNI information through CNIConfig")
+		if _, ok := validCNIs[network.CNI]; !ok {
+			return fmt.Errorf("cni %s not supported", network.CNI)
+		}
+		return nil
+	}
+	return validateCNIConfig(network.CNIConfig)
+}
+
+func validateCNIConfig(cniConfig *CNIConfig) error {
+	if cniConfig == nil {
+		return fmt.Errorf("cni not specified")
+	}
+	var cniPluginSpecified bool
+
+	if cniConfig.Cilium != nil {
+		cniPluginSpecified = true
+	}
+
+	if cniConfig.Kindnetd != nil {
+		if cniPluginSpecified {
```

Comment:
nit: you can make `cniPluginSpecified` a counter and just return error at the end if it's more than 1

That avoids nested `if` and allows to add more CNIs easily